### PR TITLE
refactor: use v2 device-sdk

### LINKED ISSUE #214: Consume V2 Device SDK (ModBus)
1. Update go.mod file to reference new v2.0.x version of module
2. Update all import paths for the module to include /v2 in the path.
3. Update Attribution.txt to add /v2 to the module reference.

See the PR for edgex-go for example:  https://github.com/edgexfoundry/edgex-go/pull/3032/files

### PR DESCRIPTION
**Pull request by hahattan**
Signed-off-by: Chris Hung <chris@iotechsys.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-modbus-go/blob/master/.github/Contributing.md

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

File: internal/driver/constant.go [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-// Copyright (C) 2018-2019 IOTech Ltd
+// Copyright (C) 2018-2021 IOTech Ltd
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package driver
 
-import sdkModel "github.com/edgexfoundry/device-sdk-go/pkg/models"
+import (
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/v2"
```
Comment (lenny-intel) [inline internal/driver/constant.go:10]: The `v2` namespace will go away. So to reduce future changes for when that happens, might consider aliasing this to what we expect the namespace will be. Same in other files.

File: internal/driver/deviceclient.go [hunk 2]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
-	case sdkModel.Uint16:
+	case v2.ValueTypeUint16:
 		var res = binary.BigEndian.Uint16(dataBytes)
-		result, err = sdkModel.NewUint16Value(req.DeviceResourceName, resTime, res)
-	case sdkModel.Uint32:
+		result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeUint16, res)
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:120]: looks like this block of repeated code can now go after the switch if `res` is define before the stitch as: `var res interface{}`. may sure res is set/cast appropriately in switch code to allow this.

File: internal/driver/deviceclient.go [hunk 3]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
 			var res = int16(binary.BigEndian.Uint16(dataBytes))
-			result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, float32(res))
-			driver.Logger.Debug(fmt.Sprintf("According to the rawType %s and the value type %s, convert integer %d to float %v ", INT16, FLOAT32, res, result.ValueToString(contract.ENotation)))
-		case sdkModel.Uint16:
+			result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeFloat32, float32(res))
+			if err != nil {
+				return nil, err
+			}
+			result.Origin = resTime
+			driver.Logger.Debugf("According to the rawType %s and the value type %s, convert integer %d to float %v ", INT16, FLOAT32, res, result.ValueToString())
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:173]: Why is this here?? The value is `int16` on line 167??

File: internal/driver/deviceclient.go [hunk 4]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
+		case v2.ValueTypeUint16:
 			var res = binary.BigEndian.Uint16(dataBytes)
-			result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, float32(res))
-			driver.Logger.Debug(fmt.Sprintf("According to the rawType %s and the value type %s, convert integer %d to float %v ", UINT16, FLOAT32, res, result.ValueToString(contract.ENotation)))
+			result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeFloat32, float32(res))
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:176]: Why are we converting the `Uint16` to a `float32`??

File: internal/driver/deviceclient.go [hunk 5]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
+			if err != nil {
+				return nil, err
+			}
+			result.Origin = resTime
+		case v2.ValueTypeInt16:
 			var res = int16(binary.BigEndian.Uint16(dataBytes))
-			result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, float32(res))
-			driver.Logger.Debug(fmt.Sprintf("According to the rawType %s and the value type %s, convert integer %d to float %v ", INT16, FLOAT32, res, result.ValueToString(contract.ENotation)))
-		case sdkModel.Uint16:
+			result, err = models.NewCommandValue(req.DeviceResourceName, v2.ValueTypeFloat32, float32(res))
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:168]: Why float32?? If really needed, then use float32 cast on line 194 so this block can be below switch. Same with similar code below.

File: internal/driver/deviceclient.go [hunk 6]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
+		if err != nil {
+			return nil, err
+		}
+		result.Origin = resTime
+	case v2.ValueTypeFloat32:
 		switch commandInfo.RawType {
-		case sdkModel.Float32:
+		case v2.ValueTypeFloat32:
 			var res = binary.BigEndian.Uint32(swap32BitDataBytes(dataBytes, commandInfo.IsByteSwap, commandInfo.IsWordSwap))
 			var floatResult = math.Float32frombits(res)
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:160]: Just set `res` rather then new var.

File: internal/driver/deviceclient.go [hunk 7]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
-	driver.Logger.Debug(fmt.Sprintf("Transfer dataBytes to CommandValue(%v) successful.", result.ValueToString(contract.ENotation)))
-	return result, err
+
+	driver.Logger.Debugf("Transfer dataBytes to CommandValue(%v) successful.", result.ValueToString())
+	return result, nil
 }
 
-func TransformCommandValueToDataBytes(commandInfo *CommandInfo, value *sdkModel.CommandValue) ([]byte, error) {
-	var byteCount = calculateByteCount(commandInfo)
+func TransformCommandValueToDataBytes(commandInfo *CommandInfo, value *models.CommandValue) ([]byte, error) {
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:229]: Add comment for func explaining why do we need to do this?

File: internal/driver/deviceclient.go [hunk 8]
```diff
@@ -105,126 +105,186 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
 	return length
 }
 
-func TransformDataBytesToResult(req *sdkModel.CommandRequest, dataBytes []byte, commandInfo *CommandInfo) (*sdkModel.CommandValue, error) {
-	var result = &sdkModel.CommandValue{}
+func TransformDataBytesToResult(req *models.CommandRequest, dataBytes []byte, commandInfo *CommandInfo) (*models.CommandValue, error) {
```
Comment (lenny-intel) [inline internal/driver/deviceclient.go:108]: Func shoud have godoc style comment

File: snap/snapcraft.yaml [hunk 9]
```diff
@@ -82,7 +82,11 @@ parts:
   device-modbus:
     source: .
     plugin: make
-    build-packages: [git]
+    build-packages:
```
Comment (lenny-intel) [inline snap/snapcraft.yaml:85]: ASC has this also:
https://github.com/edgexfoundry/app-service-configurable/blob/master/snap/snapcraft.yaml#L75
@tonyespy should this snap also have this?

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: snap/snapcraft.yaml (hunk 9 above)
Comment (tonyespy) [reply snap/snapcraft.yaml:85]: @lenny-intel yes

File: cmd/res/devices/devices.toml.example [hunk 10]
```diff
@@ -1,51 +1,7 @@
-[Writable]
```
[the remark this answers is not in the document]
Comment (tonyespy) [reply cmd/res/devices/devices.toml.example:1]: I agree that this is inconsistent. Prior to this PR we had two configuration.toml files, one in /res, and one in /res/examples. This PR takes the latter file, strips everything but the `DeviceList` (which the top-level TOML file lacks) and re-names `devices.toml.example`. The former device profile however is moved into a new directory `/res/profile`, but is left as `.yml` file, so it will be imported on startup. I think we should be consistent and either have both use a `.example` suffix to prevent them from being auto-imported, or don't add the suffix and have both files processed by default (which means the service will inherit the devices as defined in the `devices.toml` file.

**Review by tonyespy — COMMENTED** [review]
See comments inline...

File: internal/driver/deviceclient.go (hunk 5 above)
Comment (weichou1229) [reply internal/driver/deviceclient.go:168]: > Why float32??

We provide rawType for converting **integer**  to **float** value , see https://docs.edgexfoundry.org/1.3/examples/Ch-ExamplesModbusdatatypeconversion/#supported-transformations

File: internal/driver/deviceclient.go (hunk 4 above)
Comment (weichou1229) [reply internal/driver/deviceclient.go:176]: For example, the real device store 3210 as temperature value.  Then we need to divide 3210 by 100 and show the 32.1 as the actual temperature value.
see https://github.com/edgexfoundry/device-modbus-go/issues/78

File: internal/driver/deviceclient.go (hunk 8 above)
Comment (weichou1229) [reply internal/driver/deviceclient.go:108]: @hahattan Please help to add the following comment:
TransformDataBytesToResult is used to transform the device's binary data to the specified value type as the actual result.

File: internal/driver/deviceclient.go (hunk 3 above)
Comment (weichou1229) [reply internal/driver/deviceclient.go:173]: To implement rawType feature, we read the data as Int16 and convert Int16 to float32.

File: internal/driver/deviceclient.go (hunk 7 above)
Comment (weichou1229) [reply internal/driver/deviceclient.go:229]: @hahattan Please help to add the following comment:
TransformCommandValueToDataBytes transforms the reading value to binary data which is used to transfer data via Modbus protocol.

File: cmd/res/devices/devices.toml.example (hunk 10 above)
[the remark this answers is not in the document]
Comment (hahattan) [reply cmd/res/devices/devices.toml.example:1]: you're right it's inconsistent in this PR (profiles are auto-imported while devices are not).

I think we can remove the .example suffix and update the `ProfilesDir` and `DevicesDir` to just `./res`. If user want to import the example devices/profiles, they can env overwrite those 2 configuration.

@lenny-intel @tonyespy thought on this solution ?

File: internal/driver/constant.go (hunk 1 above)
Comment (hahattan) [reply internal/driver/constant.go:10]: I'm really not sure where this constants.go file would end up being, in v1 we put it in `clients` package but I doubt there will be import cycle in v2 if we do so. 

If we eventually still needs to update the import path (remove the `v2` namespace), I would prefer make the change till then.

**Comment by hahattan** [comment]
noticed that this depends on https://github.com/edgexfoundry/device-sdk-go/pull/870 to correctly use the example profile

File: cmd/res/devices/devices.toml.example (hunk 10 above)
[the remark this answers is not in the document]
Comment (iain-anderson) [reply cmd/res/devices/devices.toml.example]: Seems ok to me. I'd be tempted to rename devices.toml to modbus.test.devices.toml for clarity and consistency with the profile yml

**Review by lenny-intel — APPROVED** [review]
LGTM

File: cmd/res/configuration.toml [hunk 11]
```diff
@@ -39,6 +75,7 @@ Type = 'consul'
   MaxCmdValueLen = 256
   RemoveCmd = ''
   RemoveCmdArgs = ''
-  ProfilesDir = './res'
+  ProfilesDir = ''
+  DevicesDir = ''
   UpdateLastConnected = false
```
Comment (cloudxxx8) [inline cmd/res/configuration.toml:81]: remove the additional space

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]
I agree with the example change, but please add a statement in the README.MD to tell the user that the `ProfileDir` and `DeviceDir` in the configuration.toml is empty string by default, and they can fill './res/profiles' and './res/devices' to use the example pre-defined Device and Profile.

**Comment by cloudxxx8** [comment]
please also address Iain's comment
rename devices.toml to modbus.test.devices.toml

**Comment by cloudxxx8** [comment]
> I agree with the example change, but please add a statement in the README.MD to tell the user that the `ProfileDir` and `DeviceDir` in the configuration.toml is empty string by default, and they can fill './res/profiles' and './res/devices' to use the example pre-defined Device and Profile.

Please add a section in README, after the **Usage** section.

### Example Profile and Device
The `ProfileDir` and `DeviceDir` in the configuration.toml are empty string by default.  To use the example Profile and Device in this repository, please fill './res/profiles' and './res/devices' to `ProfileDir` and `DeviceDir` respectively.  modbus.test.device.profile.yml and modbus.test.devices.toml will be loaded and created when the Device Service starts up.  Users can modify those files or add additional Profile YAML or Device TOML to meet their needs.

**Comment by hahattan** [comment]
README updated.

**Review by cloudxxx8 — APPROVED** [review]
LGTM
